sklearn/datasets/tests/test_base.py

```diff
@@ -318,3 +321,33 @@ def test_bunch_dir():
     # check that dir (important for autocomplete) shows attributes
     data = load_iris()
     assert "data" in dir(data)
+
+
+# FIXME: to be removed in 1.2
+def test_load_boston_warning():
+    """Check that we raise the ethical warning when loading `load_boston`."""
+    warn_msg = "The Boston housing prices dataset has an ethical problem"
+    with pytest.warns(FutureWarning, match=warn_msg):
+        load_boston()
+
+
+@pytest.mark.filterwarnings("ignore:Function load_boston is deprecated")
+def test_load_boston_alternative():
+    pytest.importorskip("pandas")
+    if not os.environ.get("SKLEARN_SKIP_NETWORK_TESTS", "1") == "1":
+        raise SkipTest(
+            "This test requires an internet connection to fetch the dataset."
+        )
+    boston_openml = fetch_openml(name="boston", version=1, as_frame=True)
+    boston_sklearn = load_boston()
+
+    np.testing.assert_array_equal(
+        boston_openml.feature_names, boston_sklearn.feature_names
+    )
+    np.testing.assert_allclose(
+        boston_openml.data.to_numpy().astype(np.float64),
```

Comment:
I added this test to make sure of the consistency. It works when we get the dataframe.

However, when getting a NumPy array `as_frame=False` then we don't have the equality.
We should check if there is a bug on the `fetch_openml` because it should not happen.

Reply:
OK, so we have an issue here:

```python
In [1]: boston_sklearn.data[:, boston_sklearn.feature_names.tolist().index("RAD")]
Out[1]: 
array([ 1.,  2.,  2.,  3.,  3.,  3.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,  5.,  5.,  5.,
        3.,  3.,  3.,  3.,  3.,  3.,  3.,  3.,  3.,  3.,  3.,  4.,  4.,
        4.,  4.,  3.,  5.,  2.,  5.,  8.,  8.,  8.,  8.,  8.,  8.,  3.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,  5.,  5.,  5.,
        5.,  5.,  4.,  4.,  4.,  4.,  3.,  3.,  3.,  3.,  2.,  2.,  2.,
        2.,  4.,  4.,  4.,  2.,  2.,  2.,  2.,  2.,  5.,  5.,  5.,  5.,
        5.,  5.,  5.,  5.,  5.,  5.,  5.,  6.,  6.,  6.,  6.,  6.,  6.,
        6.,  6.,  6.,  2.,  2.,  2.,  2.,  2.,  2.,  2.,  4.,  4.,  4.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,
        5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,
        5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,
        5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  3.,  3.,  3.,
        3.,  3.,  3.,  3.,  3.,  5.,  5.,  5.,  5.,  5.,  5.,  1.,  1.,
        4.,  2.,  2.,  2.,  3.,  3.,  2.,  2.,  4.,  4.,  4.,  4.,  4.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,  5.,  5.,  5.,  8.,
        8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,  8.,
        8.,  8.,  8.,  8.,  6.,  6.,  6.,  6.,  6.,  6.,  7.,  7.,  7.,
        7.,  7.,  7.,  7.,  7.,  7.,  7.,  1.,  1.,  3.,  5.,  5.,  5.,
        5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  5.,  3.,  3.,  3.,  3.,
        3.,  4.,  4.,  4.,  4.,  4.,  5.,  5.,  5.,  5.,  1.,  1.,  1.,
        1.,  6.,  6.,  6.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,
        5.,  5.,  7.,  7.,  7.,  7.,  7.,  7.,  7.,  4.,  4.,  4.,  4.,
        4.,  4.,  4.,  4.,  4.,  4.,  4.,  4.,  5.,  5.,  5.,  5.,  5.,
        5.,  5.,  5.,  4.,  4.,  4.,  1.,  1.,  5.,  5.,  5.,  5.,  5.,
        5.,  5.,  5.,  1.,  1.,  5.,  5.,  3.,  3.,  4.,  4.,  1.,  1.,
        4.,  4.,  5.,  4.,  4., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24., 24.,
       24., 24., 24., 24., 24., 24., 24.,  4.,  4.,  4.,  4.,  4.,  6.,
        6.,  6.,  6.,  6.,  6.,  6.,  6.,  1.,  1.,  1.,  1.,  1.])

In [2]: boston_openml.data[:, boston_openml.feature_names.index("RAD")]
Out[2]: 
array([0., 1., 1., 2., 2., 2., 4., 4., 4., 4., 4., 4., 4., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 4., 4., 4., 4., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 2., 3.,
       3., 3., 3., 2., 4., 1., 4., 7., 7., 7., 7., 7., 7., 2., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 4., 4., 4., 4., 4., 4., 3., 3., 3., 3., 2.,
       2., 2., 2., 1., 1., 1., 1., 3., 3., 3., 1., 1., 1., 1., 1., 4., 4.,
       4., 4., 4., 4., 4., 4., 4., 4., 4., 5., 5., 5., 5., 5., 5., 5., 5.,
       5., 1., 1., 1., 1., 1., 1., 1., 3., 3., 3., 3., 3., 3., 3., 3., 3.,
       3., 3., 3., 3., 3., 3., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,
       4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,
       4., 4., 4., 4., 4., 4., 4., 4., 4., 2., 2., 2., 2., 2., 2., 2., 2.,
       4., 4., 4., 4., 4., 4., 0., 0., 3., 1., 1., 1., 2., 2., 1., 1., 3.,
       3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 4., 4., 4., 4., 7.,
       7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7., 7.,
       5., 5., 5., 5., 5., 5., 6., 6., 6., 6., 6., 6., 6., 6., 6., 6., 0.,
       0., 2., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 2., 2., 2.,
       2., 2., 3., 3., 3., 3., 3., 4., 4., 4., 4., 0., 0., 0., 0., 5., 5.,
       5., 3., 3., 3., 3., 3., 3., 3., 3., 4., 4., 4., 6., 6., 6., 6., 6.,
       6., 6., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 3., 4., 4., 4.,
       4., 4., 4., 4., 4., 3., 3., 3., 0., 0., 4., 4., 4., 4., 4., 4., 4.,
       4., 0., 0., 4., 4., 2., 2., 3., 3., 0., 0., 3., 3., 4., 3., 3., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8.,
       8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 8., 3., 3., 3., 3., 3.,
       5., 5., 5., 5., 5., 5., 5., 5., 0., 0., 0., 0., 0.])
```

It seems that the `fetch_opeml` with `as_frame=False` does not replace the categorical values with the one stored in the ARFF header.

Reply:
Hum, please open a dedicated issue!